src/toxcore/dht/packet/crypto_handshake.rs

```diff
@@ -0,0 +1,203 @@
+/*
+    Copyright (C) 2013 Tox project All Rights Reserved.
+    Copyright © 2016-2017 Zetok Zalbavar <zexavexxe@gmail.com>
+    Copyright © 2018 Namsoo CHO <nscho66@gmail.com>
+    Copyright © 2018 Evgeny Kurnevsky <kurnevsky@gmail.com>
+    Copyright © 2018 Roman Proskuryakov <humbug@deeptown.org>
+
+    This file is part of Tox.
+
+    Tox is libre software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    Tox is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Tox.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+/*! CryptoHandshake packet
+*/
+
+use toxcore::binary_io::*;
+use toxcore::crypto_core::*;
+use toxcore::dht::packet::cookie::EncryptedCookie;
+
+/**
+
+Serialized form:
+
+Length | Content
+------ | ------
+`1`    | `0x1a`
+`24`   | `Nonce` for the cookie
+`88`   | Cookie
+`24`   | `Nonce` for the payload
+`248`  | Payload
+
+*/
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct CryptoHandshake {
+    /// Encrypted cookie
+    pub cookie: EncryptedCookie,
+    /// Nonce for the current encrypted payload
+    pub nonce: Nonce,
+    /// Encrypted payload
+    pub payload: Vec<u8>
+}
+
+impl FromBytes for CryptoHandshake {
+    named!(from_bytes<CryptoHandshake>, do_parse!(
+        tag!("\x1a") >>
+        cookie: call!(EncryptedCookie::from_bytes) >>
+        nonce: call!(Nonce::from_bytes) >>
+        payload: take!(248) >>
+        eof!() >>
```

Comment:
why eof?

Reply:
c-toxcore checks exact length - if it's bigger then packet will be droppet. So I think we can be restrictive here too - it's better to drop suspicious packet then process them.